src/polycap-description.c

```diff
@@ -94,10 +113,161 @@ polycap_description* polycap_description_new_from_file(const char *filename)
 	}
-	fscanf(fptr,"%lf", &description->n_cap);
+	fscanf(fptr,"%" PRId64, &description->n_cap);
 	i=fgetc(fptr); //reads in \n from last line still
 	out = polycap_read_input_line(fptr);
 	fclose(fptr);
 
 	return description;
 }
+
+//===========================================
+// get a new polycap_description by providing all its properties
+polycap_description* polycap_description_new(double sig_rough, double sig_wave, double corr_length, int64_t n_cap, double d_source, double d_screen, double src_x, double src_y, double src_sigx, double src_sigy, double src_shiftx, double src_shifty, size_t nelem, int iz[], double wi[], double density, polycap_profile *profile)
+{
+	int i;
+	polycap_description *description;
+
+	//allocate some memory
+	description = malloc(sizeof(polycap_description));
+	if(description == NULL){
+		printf("Could not allocate description memory.\n");
+		exit(1);
+	}
+	description->iz = malloc(sizeof(int)*nelem);
+	if(description->iz == NULL){
+		printf("Could not allocate description->iz memory.\n");
+		exit(1);
+	}
+	description->wi = malloc(sizeof(double)*nelem);
+	if(description->wi == NULL){
+		printf("Could not allocate description->wi memory.\n");
+		exit(1);
+	}
+
+	//copy data into description structure
+	description->sig_rough = sig_rough;
+	description->sig_wave = sig_wave;
+	description->corr_length = corr_length;
+	description->n_cap = n_cap;
+	description->d_source = d_source;
+	description->d_screen = d_screen;
+	description->src_x = src_x;
+	description->src_y = src_y;
+	description->src_sigx = src_sigx;
+	description->src_sigy = src_sigy;
+	description->src_shiftx = src_shiftx;
+	description->src_shifty = src_shifty;
+	description->nelem = nelem;
+	description->density = density;
+	for(i=0; i<description->nelem; i++){
+		description->iz[i] = iz[i];
+		description->wi[i] = wi[i]; //assumes weights are already provided as fractions (not percentages)
+	}
+
+	//allocate profile memory
+	description->profile = malloc(sizeof(polycap_profile));
+	if(description->profile == NULL){
+		printf("Could not allocate description->profile memory.\n");
+		exit(1);
+	}
+	description->profile->z = malloc(sizeof(double)*(profile->nmax+1));
+	if(description->profile->z == NULL){
+		printf("Could not allocate description->profile->z memory.\n");
+		exit(1);
+	}
+	description->profile->cap = malloc(sizeof(double)*(profile->nmax+1));
+	if(description->profile->cap == NULL){
+		printf("Could not allocate description->profile->cap memory.\n");
+		exit(1);
+	}
+	description->profile->ext = malloc(sizeof(double)*(profile->nmax+1));
+	if(description->profile->ext == NULL){
+		printf("Could not allocate description->profile->ext memory.\n");
+		exit(1);
+	}
+	
+	// copy description->profile values to profile
+	for(i=0;i<=profile->nmax;i++){
+		description->profile->z[i] = profile->z[i];
+		description->profile->cap[i] = profile->cap[i];
+		description->profile->ext[i] = profile->ext[i];
```

Comment:
is `description->profile->ext[i]` de juiste methode om de waarde van ext in structure pointer profile in structure pointer description aan te passen? Ik denk van wel als ik erover nadenk, maar ben ook weer geen held met pointers, laat staan pointers in pointers (pointerception?)

Reply:
Hmmm wat je hier doet is een volledige kopie maken van `profile`. Is dat nodig? Kan je niet even goed de originele `profile` gebruiken? In dat geval kan je gewoon schrijven:

````c
description->profile = profile;
````

ipv lijnen 169-194.

Dit betekent natuurlijk wel dat enige wijzigingen in de originele `profile` ook zichtbaar zullen zijn in de `description`, en je kan ook niet langer `description->profile` free'en in `polycap_description_free`

Reply:
Welja, khad erover getwijfeld. Gewoon de pointer doorverwezen (wat ik eerst had gedaan) leek me idd het makkelijkste, maar was bang dat je dan eventueel later eens profile zou free'en, waarop description ook de info kwijt is. Wist niet wat het beste was...
Aan jou de keus :) Als je denkt dat volledige kopie te veel hassle is (of gewoon voor memory sake) dan sluis ik wel gewoon de pointer door :)

Reply:
Wel als je schrik hebt dat de gebruiker later nog met `profile` zou foefelen is de deep copy wel het beste dat je kan doen.

Geheugen moet je je niet aantrekken bij zulke kleine arrays, we leven niet meer in de vroege jaren 90 😄